core/store/src/trie/trie_storage.rs

```diff
@@ -208,3 +333,111 @@ impl TouchedNodesCounter {
+        );
+    }
+
+    #[test]
+    fn test_pop() {
+        let mut trie_cache = TrieCache::new(5);
+        let key = hash(&[0, 1, 2]);
+        let value = vec![1u8];
+
+        trie_cache.put(key, &value);
+        assert_eq!(
+            trie_cache.get_with_cost(&key),
+            RawBytesWithCost { value: Some(value), cost: TrieNodeRetrievalCost::Full }
+        );
+
+        trie_cache.pop(&key);
+        assert_matches!(trie_cache.get_with_cost(&key), RawBytesWithCost { value: None, .. });
+    }
+
+    #[test]
+    fn test_shard_cache_size() {
+        let shard_cache_size = 5;
+        let mut trie_cache = TrieCache::new(shard_cache_size);
+
+        (0..shard_cache_size as u8 + 1).for_each(|i| trie_cache.put(hash(&[i]), &[i]));
+
+        assert_matches!(
+            trie_cache.get_with_cost(&hash(&[0u8])),
+            RawBytesWithCost { value: None, .. }
+        );
+        (1..shard_cache_size as u8 + 1).for_each(|i| {
+            let value = vec![i];
+            assert_eq!(
+                trie_cache.get_with_cost(&hash(&value)),
+                RawBytesWithCost { value: Some(value), cost: TrieNodeRetrievalCost::Full }
+            )
+        });
+    }
+
+    #[test]
+    fn test_positions_and_costs() {
+        let mut trie_cache = TrieCache::new(5);
+        let value = vec![1u8];
+        let key = hash(&value);
+
+        assert_matches!(trie_cache.cache_state, TrieCacheState::CachingShard);
+        assert_matches!(trie_cache.get_cache_position(&key), CachePosition::None);
+
+        trie_cache.put(key, &value);
+        assert_matches!(trie_cache.get_cache_position(&key), CachePosition::ShardCache(_));
+
+        trie_cache.cache_state = TrieCacheState::CachingChunk;
+        assert_matches!(trie_cache.get_cache_position(&key), CachePosition::ShardCache(_));
+        assert_eq!(
+            trie_cache.get_with_cost(&key),
+            RawBytesWithCost { value: Some(value.clone()), cost: TrieNodeRetrievalCost::Full }
+        );
+        assert_matches!(trie_cache.get_cache_position(&key), CachePosition::ChunkCache(_));
+        assert_eq!(
+            trie_cache.get_with_cost(&key),
+            RawBytesWithCost { value: Some(value), cost: TrieNodeRetrievalCost::Free }
+        );
+
+        trie_cache.pop(&key);
+        assert_matches!(trie_cache.get_cache_position(&key), CachePosition::None);
+    }
+
+    #[test]
+    fn test_replacement() {
+        let mut trie_cache = TrieCache::new(5);
+        let value = vec![1u8];
+        let key = hash(&value);
+
+        trie_cache.put(key, &value);
+        assert_eq!(trie_cache.shard_cache.get(&key), Some(&value));
+        assert!(!trie_cache.chunk_cache.contains_key(&key));
+
+        trie_cache.cache_state = TrieCacheState::CachingChunk;
+        let value = vec![2u8];
+        trie_cache.put(key, &value);
+        assert!(!trie_cache.shard_cache.contains(&key));
+        assert_eq!(trie_cache.chunk_cache.get(&key), Some(&value));
+    }
```

Comment:
we might have a nasty bug here...

trie.push(key, small_value)
assert trie.get(key) == small_value
trie.push(key, large_value) 
assert trie.get(key) == ??? // I think that current code will return small_value if "large_value" is > TRIE_MAX_SHARD_CACHE_SIZE